receiver/dockerstatsreceiver/receiver.go

```diff
@@ -0,0 +1,182 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerstatsreceiver
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/redisreceiver/interval"
+)
+
+var _ component.MetricsReceiver = (*Receiver)(nil)
+
+type Receiver struct {
+	config       *Config
+	logger       *zap.Logger
+	nextConsumer consumer.MetricsConsumerOld
+	client       *DockerClient
+	runner       *interval.Runner
+	obsCtx       context.Context
+	runnerCancel context.CancelFunc
+	transport    string
+}
+
+func NewReceiver(
+	_ context.Context,
+	config *Config,
+	logger *zap.Logger,
+	nextConsumer consumer.MetricsConsumerOld,
+) (component.MetricsReceiver, error) {
+	err := config.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	parsed, err := url.Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine receiver transport: %w", err)
+	}
+
+	receiver := Receiver{
+		config:       config,
+		nextConsumer: nextConsumer,
+		logger:       logger,
+		transport:    parsed.Scheme,
+	}
+
+	return &receiver, nil
+}
+
+func (r *Receiver) Start(ctx context.Context, host component.Host) error {
+	var err error
+	r.client, err = NewDockerClient(r.config, r.logger)
+	if err != nil {
+		return err
+	}
+
+	r.obsCtx = obsreport.ReceiverContext(ctx, typeStr, r.transport, r.config.Name())
+
+	runnableCtx, cancel := context.WithCancel(ctx)
+
+	runnable := newRunnable(runnableCtx, cancel, r)
+	r.runner = interval.NewRunner(r.config.CollectionInterval, runnable)
+	r.runnerCancel = cancel
+
+	go func() {
+		if err := r.runner.Start(); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
+	return nil
+}
+
+func (r *Receiver) Shutdown(ctx context.Context) error {
+	r.runnerCancel()
+	r.runner.Stop()
+	return nil
+}
+
+type Runnable struct {
+	ctx               context.Context
+	cancel            func()
+	receiver          *Receiver
+	successfullySetup bool
+}
+
+var _ interval.Runnable = (*Runnable)(nil)
+
+func newRunnable(
+	ctx context.Context,
+	cancel func(),
+	receiver *Receiver,
+) *Runnable {
+	return &Runnable{
+		ctx:      ctx,
+		cancel:   cancel,
+		receiver: receiver,
+	}
+}
+
+func (r *Runnable) Setup() error {
+	err := r.receiver.client.StartWatchingContainers(r.ctx)
```

Comment:
There's a thing called the Law of Demeter which says that you're not supposed to call a method on a member variable of a member variable. There's a simple fix though, where you can just create a `StartWatchingContainers` method on the receiver and have that method call `client.StartWatchingContainers(ctx)`. I believe this helps reduce coupling and improve encapsulation. There are a couple more opportunities to do this below as well.

Reply:
I consider the runnable part of the receiver and think the proxying would be a bit overly verbose.  How about removing the runnable intermediary and have the receiver implement its interface?

Reply:
IMO that would be preferable to this, but you also have to be careful about making sure each class just Does One Thing. :-) Up to you unless others have an opinion.

Reply:
Updated